### PR TITLE
add nudging

### PR DESCRIPTION
**Pull request by fywalter**
Nudging implementation without kv cache, support models from different families (tested on `Llama-3.1-Tulu-3-8B` (base) + `Gemma-2-2b-it` / `Llama-3.1-Tulu-3-8B-DPO` (nudging)) and batch size > 1.

Remark: not sure which category (api, text, weight, or logits) does nudging belong to. Using API currently.

### CODE REVIEW

**Comment by BunsenFeng** [comment]
Thank you, Yu! API-level sounds good, and your PR looks great! Yuyang and I will begin testing on our end and let you know if any issues.

**Comment by BunsenFeng** [comment]
I made a minor change to the order of paragraphs in `method/user_readme.md`.

**Review by leopoldwhite — APPROVED** [review]
successfully ran the three configs.

**Comment by BunsenFeng** [comment]
Amazing! I will test a config or two (on Fri?) and we should be good.

**Comment by BunsenFeng** [comment]
Thank you, Yu! @fywalter

1. It seems like `len(model_names) > 2` is supported? If yes, in the readme, maybe `- nudging_model_id (default 1): the index of the nudging model in model_names.` should change to something like `anything not the base is a potential nudging model`, if that's the case?
2. Seems like the GPU vram usage is really high: I'm running `len(model_names) == 3` with 7-8B LMs with 3 140GB H200 GPUs and batch size 16, still OOM? Changing to 4 works. Would you have any idea why VRAM usage is high?

Apart from these, I ran config4 in [here](https://docs.google.com/document/d/1kVsL7tB1t4RMZXo9Bcuve6DI2tmrJXoiiODTEFGqDbg/edit?usp=sharing) and it successfully finished.

**Comment by fywalter** [comment]
Hi Shangbin,

1. `len(model_names) > 2` is supported, I will change readme accordingly.
2. Looking into it, likely because I am not using `torch.no_grad()` or `torch.inference_mode()`. Will let you know what I find.

**Comment by fywalter** [comment]
The memory issue is indeed caused by not applying inference_mode. However, due to not using kv_cache (I am following the `logic_arithmatic.py`), the inference time for `batchsize = 4` is roughly 4x the longest inference time for this batch with `batchsize = 1` if all the queries in the batch do not complete. I think we need to be aware of this before we have the kv-cached version implemented, which I guess is not the priority at this point?
And I guess a simple fix is to group the samples in the datasets by their query lengths.

Also noticed a potential bug for the `logic_arithmatic.py`
```python
# line 87 -89 in generate
        for step in tqdm(range(max_new_tokens)):
            inputs = {'input_ids': input_ids}
            list_of_outputs = self.forward(inputs, return_dict=True) 
```
Currently we only pass the input_ids to the model without attention_mask, could this be problematic for batched generation?

**Comment by fywalter** [comment]
1. updated readme
2. added `torch.inforence_mode()`, so large batch size should be good
3. added dynamic batching to avoid forwarding terminated samples in the batch
4. fixed attention mask bug

**Comment by BunsenFeng** [comment]
I pushed a minor change from ">2" to ">=2".

Confirming the current branch works. Thank you for your suggestions on kv caching and on logit_arithmetic: will look into them and open separate PRs.

**Review by BunsenFeng — APPROVED** [review]
lgtm
